<img style="float: left; margin: 30px 15px 15px 15px;" src="https://pngimage.net/wp-content/uploads/2018/06/logo-iteso-png-5.png" width="300" height="500" /> 
    
    
### <font color='navy'> Simulación de procesos financieros. 

**Nombres:** Moisés Flores Ortiz e Ivette Landaverde Mercado
              

**Fecha:** 15 de febrero del 2021.

**Expediente** : Moisés Flores Ortiz: 722183. Ivette Landaverde Mercado: 722626.
**Profesor:** Oscar David Jaramillo Zuluaga.

# Tarea 4: Clase 7

### Link de Github
[Github](https://github.com/MoyMFO/ProyectoConjunto_MFlores_ILandaverde)

In [1]:
import random
import numpy as np
import matplotlib.pyplot as plt 
from functools import reduce
import time

# Tarea

# 1
Como ejemplo simple de una simulación de Monte Carlo, considere calcular la probabilidad de una suma particular del lanzamiento de tres dados (cada dado tiene valores del uno al seis). Además cada dado tiene las siguientes carácterísticas: el primer dado no está cargado (distribución uniforme todos son equiprobables); el segundo y tercer dado están cargados basados en una distribución binomial con parámetros (`n=5, p=0.5` y `n=5, p=0.2`). Calcule la probabilidad de que la suma resultante sea 7, 14 o 18. 

In [2]:
# Solución


### Solución 1: Moisés Flores

In [3]:
# Solución
#Imoprtamos librerias
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import os
import scipy.stats as st
import seaborn as sns; sns.set()
#Dado 1 con distribución normal uniforme
dado_1 = np.random.randint(1,7,100000)
#Dado 2 cargado en base a distribución binomial con n=5 y p=0.5
dado_2 = st.binom(n=5,p=0.5,loc=1).rvs(size=100000)
#Dado 3 cargado en base a distribución binomial con n=5, p=0.2
dado_3 = st.binom(n=5,p=0.2,loc=1).rvs(size=100000)

In [4]:
#En el vector 'resultados_suma' se encuentran los resultados de cada juego
resultados_suma = dado_1 + dado_2 + dado_3
#Para la probabilidad que la suma resultante sea 7
resultados_7 = len(resultados_suma[resultados_suma==7])/len(resultados_suma)
print(f'la probabilidad de que la suma sea 7: {resultados_7}')
#Para la probabilidad que la suma resultante sea 14
resultados_14 = len(resultados_suma[resultados_suma==14])/len(resultados_suma)
print(f'la probabilidad de que la suma sea 14: {resultados_14}')
#Para la probabilidad que la suma resultante sea 18
resultados_18 = len(resultados_suma[resultados_suma==18])/len(resultados_suma)
print(f'la probabilidad de que la suma sea 18: {float(resultados_18)}')

la probabilidad de que la suma sea 7: 0.12778
la probabilidad de que la suma sea 14: 0.01373
la probabilidad de que la suma sea 18: 0.0


### Respuesta 
 La respuesta es:
> $\text{La probabilidad de que la suma sea 7:  0.12757}$

> $\text{La probabilidad de que la suma sea 14: 0.01385}$

> $\text{La probabilidad de que la suma sea 18: 1e-05}$

$$\text{Con estos dados lo menos probable es que la suma de 18 ($\%$ a la larga), mientras que solo hay el 1.3$\%$ de}$$  $$\text{probabilidad de obtener 14 y se cuenta con el 12.7$\%$ de probabilidad de obtener 7, si solo hay estas opciones lo mejor es jugar poor el 7.}$$

### Solución 1: Ivette Landaverde

In [1]:
###### Solución
import random
import numpy as np
def tiro_triple(n):
    
    tiros=np.zeros(n)
    
    def llenado(i):
        
        d1=np.random.choice(range(1,7))
        d2=np.random.binomial(5, 0.5, 1)
        d3=np.random.binomial(5, 0.2, 1)
        
        tiros[i]=d1+d2+d3
    
    [llenado(i) for i in range(n)]
    
    return tiros

n=10000

escenarios=tiro_triple(n)

suma_7=[x for x in escenarios if x==7]
suma_14=[x for x in escenarios if x==14]
suma_18=[x for x in escenarios if x==18]

prob_7=len(suma_7)/n
prob_14=len(suma_14)/n
prob_18=len(suma_18)/n


prob_7, prob_14, prob_18

(0.1605, 0.0005, 0.0)

### Respuesta 
 La respuesta es:

In [2]:
print("La probabilidad de que la suma de los dados sea 7 es:",prob_7)
print("La probabilidad de que la suma de los dados sea 14 es:",prob_14)
print("La probabilidad de que la suma de los dados sea 18 es:",prob_18)

La probabilidad de que la suma de los dados sea 7 es: 0.1605
La probabilidad de que la suma de los dados sea 14 es: 0.0005
La probabilidad de que la suma de los dados sea 18 es: 0.0


# 2 Ejercicio de aplicación- Cafetería Central

Premisas para la simulación:
- Negocio de alimentos que vende bebidas y alimentos.
- Negocio dentro del ITESO.
- Negocio en cafetería central.
- Tipo de clientes (hombres y mujeres).
- Rentabilidad del 60%.

## Objetivo
Realizar una simulación estimado el tiempo medio que se tardaran los clientes en ser atendidos entre el horario de 6:30 a 1 pm. Además saber el consumo. 
**Analizar supuestos y limitantes**

## Supuestos en simulación
Clasificación de clientes: 
- Mujer  = 1 $\longrightarrow$ aleatorio < 0.5
- Hombre = 0 $\longrightarrow$ aleatorio $\geq$ 0.5.

Condiciones iniciales:
- Todas las distrubuciones de probabilidad se supondrán uniformes.
- Tiempo de simulación: 6:30 am - 1:30pm $\longrightarrow$ T = 7 horas = 25200 seg.
- Tiempo de llegada hasta ser atendido: Min=5seg, Max=30seg.
- Tiempo que tardan los clientes en ser atendidos:
    - Mujer: Min = 1 min= 60seg, Max = 5 min = 300 seg
    - Hombre: Min = 40 seg, Max = 2 min= 120 seg
- Consumo según el tipo de cliente:
    - Mujer: Min = 30 pesos, Max = 100 pesos
    - Hombre: Min = 20 pesos, Max = 80 pesos

Responder las siguientes preguntas basados en los datos del problema:
1. ¿Cuáles fueron los gastos de los hombres y las mujeres en 5 días de trabajo?.
2. ¿Cuál fue el consumo promedio de los hombres y mujeres?
3. ¿Cuál fue el número de personas atendidas por día?
4. ¿Cuál fue el tiempo de atención promedio?
5. ¿Cuánto fue la ganancia promedio de la cafetería en 5 días de trabajo y su respectiva rentabilidad?

In [7]:
################## Datos del problema
d = 5
T =25200
T_at_min = 5; T_at_max = 30
T_mujer_min =60; T_mujer_max = 300
T_hombre_min = 40; T_hombre_max = 120
C_mujer_min = 30; C_mujer_max = 100
C_hombre_min = 20; C_hombre_max = 80


### Solución 2: Moisés Flores

In [8]:
###### Solución
###### Solución
#Primero simulamos el tiempo de atención total de un hombre
#Planteamos 10,000 escenarios
escenarios = 10000
#Se simulará el tiempo hasta ser atendido en 5 días 
tiempo_total_atencion_hombre = st.uniform(loc=(T_at_min + T_hombre_min),
                                       scale=(T_at_max + T_hombre_max) - (T_at_min + T_hombre_min) ).rvs(size=[escenarios,d])
#Obtenemos la media del tiempo hasta ser atendido por cada día
tiempo_total_atencion_hombre_media =  tiempo_total_atencion_hombre.mean(axis=0)
tiempo_total_atencion_hombre_media

array([97.43633923, 97.17383089, 97.95434801, 97.48786087, 97.8029369 ])

In [9]:
#Primero simulamos el tiempo de atención total de un mujer
#Planteamos 10,000 escenarios
escenarios = 10000
#Se simulará el tiempo hasta ser atendido en 5 días 
tiempo_total_atencion_mujer = st.uniform(loc=(T_at_min + T_mujer_min),
                                       scale=(T_at_max + T_mujer_max) - (T_at_min + T_mujer_min) ).rvs(size=[escenarios,d])
#Obtenemos la media del tiempo hasta ser atendido por cada día
tiempo_total_atencion_mujer_media =  tiempo_total_atencion_mujer.mean(axis=0)
tiempo_total_atencion_mujer_media

array([197.79237896, 197.6476922 , 198.02392944, 197.93512485,
       197.53529993])

In [10]:
#Obtenemos el tiempo promedio en que una persona (sin importar género) 
#tarda en la cafeteria por cada día, es decir, desde esperar hasta ser atendida
#como en ser atendia
tiempo_total_medio = (tiempo_total_atencion_hombre_media + tiempo_total_atencion_mujer_media)/2
tiempo_total_medio

array([147.61435909, 147.41076154, 147.98913873, 147.71149286,
       147.66911841])

In [11]:
#Ahora podemos obtener el numero de personas que fueron atendidas por día
personas_atendidas_por_dia = np.array(list(map(lambda x: T/x,tiempo_total_medio)))
np.round(personas_atendidas_por_dia)

array([171., 171., 170., 171., 171.])

In [12]:
#Primero simulamos el numero de mujeres y el numero de hombres que asistieron por día
def escenarios_genero_x_dia(asistentes_diarios):
    return st.uniform(loc=0, scale=1).rvs(size=[escenarios,asistentes_diarios])
esc_gen_x_dia_distribucion = list(map(lambda x:escenarios_genero_x_dia(int(x)).mean(axis=0),personas_atendidas_por_dia))
### Ahora obtenemos por día hombres y mujeres
vec_hombres_x_cada_día = list((map(lambda x:len(x[x>=0.5]),esc_gen_x_dia_distribucion)))
vec_mujeres_x_cada_día = list((map(lambda x:len(x[x<0.5]),esc_gen_x_dia_distribucion)))
print(f"""Estos son los hombres por día simulando 10k escenarios 
diferentes cada día:{vec_hombres_x_cada_día}""")
print(f"""Estos son las mujeres por día simulando 10k escenarios 
diferentes cada día:{vec_mujeres_x_cada_día}""")

Estos son los hombres por día simulando 10k escenarios 
diferentes cada día:[82, 74, 86, 80, 88]
Estos son las mujeres por día simulando 10k escenarios 
diferentes cada día:[88, 96, 84, 90, 82]


Ya que se calculo el numero de personas atendidas por día y su género podemos responder las preguntas

1. ¿Cuáles fueron los gastos de los hombres y las mujeres en 5 días de trabajo?.

In [13]:
### Hombres
def esc_gasto_hombre_x_dia(hombres_diarios):
    return st.uniform(loc=C_hombre_min, scale=C_hombre_max-C_hombre_min).rvs(
        size=[escenarios,hombres_diarios])
esc_gasto_H_x_dia = list(map(lambda x:esc_gasto_hombre_x_dia(int(x)).mean(axis=0),vec_hombres_x_cada_día))
##Ahora calculamos la suma del gasto de cada hombre
suma_gasto_H_x_dia = list(map(lambda x: x.sum(),esc_gasto_H_x_dia))
print(f'Esto es lo que gastaría por día: {suma_gasto_H_x_dia}')
print(f'Esto es lo que gastarían en total en los 5 días: {np.array(suma_gasto_H_x_dia).sum()}')

Esto es lo que gastaría por día: [4100.678265156535, 3699.8500601201727, 4298.086869105967, 3999.384460518998, 4400.165372767174]
Esto es lo que gastarían en total en los 5 días: 20498.165027668845


In [14]:
### Mujeres
def esc_gasto_mujer_x_dia(mujeres_diarios):
    return st.uniform(loc=C_mujer_min, scale=C_mujer_max-C_mujer_min).rvs(
        size=[escenarios,mujeres_diarios])
esc_gasto_M_x_dia = list(map(lambda x:esc_gasto_mujer_x_dia(int(x)).mean(axis=0),vec_mujeres_x_cada_día))
##Ahora calculamos la suma del gasto de cada hombre
suma_gasto_M_x_dia = list(map(lambda x: x.sum(),esc_gasto_M_x_dia))
print(f'Esto es lo que gastaría por día: {suma_gasto_M_x_dia}')
print(f'Esto es lo que gastarían en total en los 5 días: {np.array(suma_gasto_M_x_dia).sum()}')

Esto es lo que gastaría por día: [5718.356513560944, 6238.165318211821, 5461.730125431099, 5853.25305291095, 5329.3448168813175]
Esto es lo que gastarían en total en los 5 días: 28600.849826996135


2. ¿Cuál fue el consumo promedio de los hombres y mujeres?

In [15]:
#Hombres
promedio_gasto_H_x_dia = list(map(lambda x: x.mean(),esc_gasto_H_x_dia))
print(f'Esto es lo que consumen cada hombre en promedio por día: {promedio_gasto_H_x_dia}')
print(f'Esto es lo que consumen cada hombre en promedio en los 5 días: {np.array(promedio_gasto_H_x_dia).sum()}')

Esto es lo que consumen cada hombre en promedio por día: [50.00827152629921, 49.99797378540774, 49.97775429192985, 49.992305756487475, 50.00187923599061]
Esto es lo que consumen cada hombre en promedio en los 5 días: 249.97818459611486


In [16]:
#Mujeres
promedio_gasto_M_x_dia = list(map(lambda x: x.mean(),esc_gasto_M_x_dia))
print(f'Esto es lo que consumen en promedio por día: {promedio_gasto_M_x_dia}')
print(f'Esto es lo que consumen en promedio en los 5 días: {np.array(promedio_gasto_M_x_dia).sum()}')

Esto es lo que consumen en promedio por día: [64.981324017738, 64.98088873137313, 65.0205967313226, 65.03614503234388, 64.99200996196728]
Esto es lo que consumen en promedio en los 5 días: 325.0109644747449


3. ¿Cuál fue el número de personas atendidas por día?

In [17]:
# Este dato se obtuvo de dividir el tiempo durante el que está abierto el negocio
# dividido entre la suma de las simulación del tiempo medio hasta ser atendido de cada dia
# más el tiempo en ser atendido de cada día (el cálculo se encuentra al inicio)
personas_atendidas_por_día = np.array(list(map(lambda x: T/x,tiempo_total_medio)))
print(f'Número de personas atendidas por día: {np.round(personas_atendidas_por_día)}')

Número de personas atendidas por día: [171. 171. 170. 171. 171.]


4. ¿Cuál fue el tiempo de atención promedio?

In [18]:
# Este dato se obtuvo de la suma de la simulación del tiemmpo hasta ser atendido
# más la simulación del tiempo medio en ser atendido
tiempo_total_medio = (tiempo_total_atencion_hombre_media + tiempo_total_atencion_mujer_media)/2
print(f""" Tiempo promedio de atención diario: {tiempo_total_medio} y tiempo medio general en ser atendido en los 5 días: {tiempo_total_medio.mean()}""")

 Tiempo promedio de atención diario: [147.61435909 147.41076154 147.98913873 147.71149286 147.66911841] y tiempo medio general en ser atendido en los 5 días: 147.6789741276661


5. ¿Cuánto fue la ganancia promedio de la cafetería en 5 días de trabajo y su respectiva rentabilidad?

In [19]:
# La ganacia promedio de la cafeteria es el resultado de la suma de los gastos medios 
# por día tanto de hombres como de mujeres
suma_H_mas_M = np.array(suma_gasto_H_x_dia) + np.array(suma_gasto_M_x_dia)
print(f"""Ganancia promedio de cada día {suma_H_mas_M} 
      En Total:{suma_H_mas_M.sum()} """)

Ganancia promedio de cada día [9819.03477872 9938.01537833 9759.81699454 9852.63751343 9729.51018965] 
      En Total:49099.01485466498 


In [20]:
# La rentabilidad promedio de la cafeteria sería el 60% de estás sumas
print(f"""Rentabilidad promedio de cada día {suma_H_mas_M * 0.6}
      En Total:{suma_H_mas_M.sum()*0.6} """)

Ganancia promedio de cada día [5891.42086723 5962.809227   5855.89019672 5911.58250806 5837.70611379]
      En Total:29459.408912798986 


### Respuesta 

1. ¿Cuáles fueron los gastos de los hombres y las mujeres en 5 días de trabajo?.
> Hombres: Esto es lo que gastaría por día: [4100.678265156535, 3699.8500601201727, 4298.086869105967, 3999.384460518998, 4400.165372767174] Esto es lo que gastarían en total en los 5 días: 20498.165027668845
> Mujeres: Esto es lo que gastaría por día: [5718.356513560944, 6238.165318211821, 5461.730125431099, 5853.25305291095, 5329.3448168813175] Esto es lo que gastarían en total en los 5 días: 28600.849826996135
2. ¿Cuál fue el consumo promedio de los hombres y mujeres?
> Esto es lo que consumen cada hombre en promedio por día: [50.00827152629921, 49.99797378540774, 49.97775429192985, 49.992305756487475, 50.00187923599061] Esto es lo que consumen cada hombre en promedio en los 5 días: 249.97818459611486
> Esto es lo que consumen cada mujer en promedio por día: [64.981324017738, 64.98088873137313, 65.0205967313226, 65.03614503234388, 64.99200996196728]
Esto es lo que consumen en promedio en los 5 días: 325.0109644747449
3. ¿Cuál fue el número de personas atendidas por día?
> Número de personas atendidas por día: [171. 171. 170. 171. 171.]
4. ¿Cuál fue el tiempo de atención promedio?
> Tiempo promedio de atención diario: [147.61435909 147.41076154 147.98913873 147.71149286 147.66911841] y tiempo medio general en ser atendido en los 5 días: 147.6789741276661
5. ¿Cuánto fue la ganancia promedio de la cafetería en 5 días de trabajo y su respectiva rentabilidad?
> Rentabilidad promedio de cada día [5891.42086723 5962.809227   5855.89019672 5911.58250806 5837.70611379] En Total:29459.408912798986 
> 

### Solución 2: Ivette Landaverde

In [3]:
###### Solución
#Datos iniciales
d = 5
T =25200
T_at_min = 5; T_at_max = 30
T_mujer_min =60; T_mujer_max = 300
T_hombre_min = 40; T_hombre_max = 120
C_mujer_min = 30; C_mujer_max = 100
C_hombre_min = 20; C_hombre_max = 80


#Definiendo listas iniciales
# Listas
T_aten = [] #Tiempo de atención por persona
c_h = [] #consumo hombres
c_m = [] #Consumo mujeres
p_dia = []

P_aten = 0 #Personas atendidas en total
j = 0 #Indice dentro de la variable personas
m = 0 # cuántas mujeres
h = 0 #cuántos hombres
g = np.zeros(5) #Ganancias por día
def eventoClienteFemenino():
    t_llegada = np.random.randint(T_at_min,T_at_max)
    #evento mujer
    # Regresa una lista de dos cosas: tiempo de atención total y consumo
    return [np.random.randint(T_mujer_min,T_mujer_max) + t_llegada,np.random.randint(C_mujer_min,C_mujer_max)]
def eventoClienteMasculino():
    t_llegada = np.random.randint(T_at_min,T_at_max)
    #Evento hombre
    return [np.random.randint(T_hombre_min,T_hombre_max)+t_llegada,np.random.randint(C_hombre_min,C_hombre_max)]
for i in range(d):
    t = 0 #tiempo restante en el día
    loc = 0 #Variable local para trabajar
    j=0 #Personas en el día
    while t<T:
        if np.random.rand() < 0.5:
            loc = eventoClienteFemenino()
            m +=1
            c_m.append(loc[1])
        else:
            loc = eventoClienteMasculino()
            h+=1
            c_h.append(loc[1])
        t+=loc[0]
        j+=1
        T_aten.append(loc[0])
        g[i] += loc[1]*.6
    p_dia.append(j) 
        
#Respuestas del ejercicio
#1.- ¿Cuáles fueron los gastos de los hombres y las mujeres en 5 días de trabajo?.
print("Los hombres gastaron",np.array(c_h).sum(),"y las mujeres gastaron",np.array(c_m).sum())
#2.- ¿Cuál fue el consumo promedio de los hombres y mujeres?
print("Los hombres consumieron en promedio",np.array(c_h).mean(),"mientras que las mujeres consumieron en promedio",np.array(c_m).mean())
#3.-¿Cuál fue el número de personas atendidas por día?
print("Las personas atendidas por día fueron:",p_dia)
#4.-¿Cuál fue el tiempo de atención promedio?
print("El tiempo de atención promedio fue de:",np.array(T_aten).mean())
#5.-'¿Cuánto fue la ganancia promedio de la cafetería en 5 días de trabajo y su respectiva rentabilidad?
print("La ganancia promedio por dia fue de:",g.mean(),"y su rentabilidad fue del:",g.sum()/(np.array(c_h).sum()+np.array(c_m).sum()))

Los hombres gastaron 21019 y las mujeres gastaron 27622
Los hombres consumieron en promedio 49.6903073286052 mientras que las mujeres consumieron en promedio 62.92027334851936
Las personas atendidas por día fueron: [170, 169, 166, 171, 186]
El tiempo de atención promedio fue de: 146.64849187935036
La ganancia promedio por dia fue de: 5836.92 y su rentabilidad fue del: 0.6


### Respuesta 
 La respuesta es: